### DISCUSSION TITLE
Support pydantic based api

### DISCUSSION THREAD

**Discussion by sasano8** [fastapi/fastapi#8488] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Depends
from pydantic import BaseModel

app: FastAPI = PydanticbasedApi()


def get_current_user():
    return "admin"


@app.post("/user")
class CreateUser(BaseModel):
    name: str
    mail: str

    async def __call__(self, user=Depends(get_current_user)):
        return "success"
```

### Description

It is convenient to be able to declare the body like Example Code.
Would you like to support pydantic based api?

When declaring a body, it becomes more descriptive and more versatile.

### Wanted Solution

The implementation example of PydanticbasedApi is as follows.

``` python
class PydanticbasedApi(FastAPI):
    @staticmethod
    def wrap_func(func_or_cls):
        import asyncio
        import inspect
        from functools import wraps
        from typing import Callable

        if not inspect.isfunction(func_or_cls) and issubclass(func_or_cls, BaseModel):
            if not issubclass(func_or_cls, Callable):  # type: ignore
                raise TypeError("Must be Callable.")

            if asyncio.iscoroutinefunction(func_or_cls.__call__):

                @wraps(func_or_cls.__call__)
                async def wrapper(self, *args, **kwargs):
                    return await self(*args, **kwargs)

            else:

                @wraps(func_or_cls.__call__)
                def wrapper(self, *args, **kwargs):
                    return self(*args, **kwargs)

            wrapper.__name__ = func_or_cls.__name__
            wrapper.__annotations__["self"] = func_or_cls
            return wrapper

        else:
            return func_or_cls

    def post(self, *args, **kwargs):
        add_func = super().post

        def wrapped(func_or_cls):
            nonlocal add_func
            func = self.wrap_func(func_or_cls)
            add_func = add_func(*args, **kwargs)
            return add_func(func)

        return wrapped
```

### Wanted Code

```python
...
```

### Operating System

Other

### FastAPI Version

0.76.0

### Python Version

Python 3.8.12

**Comment by hellocoldworld** [comment, marked as answer]
I'm sorry, but the possible advantages of the proposed implementation are not clear to me.

Could you elaborate more on possible advantages and disadvantages of this form of use?

However, if it's just an extension to FastAPI, maybe it should be a separate package.

**Comment by sasano8** [comment]
1. Elimination of duplicate meanings

Would you like to combine CreateUser and create_user together?

```
class User(BaseModel):
  id: int
  name: str
  created_at: datetime
  
class CreateUser(BaseModel):
  name :str

@app.post("/user")
def create_user(body: CreateUser):
  return User(id=1, **body.dict(), created_at=datetime.now())

```

When diverting processing,
In this way, the function + schema can be only the schema.

``` python
result = create_user(CreateUser(name="aaa"))
```

``` python
result = CreateUser(name="aaa").__call__()
```


2. A place for to declare only one body

This doesn't really bother me, but func2 produces Body3 which contains Body1 and Body2.
The difference in the generated schema depending on the type and number of body may cause a bad effect in cooperation with a third party.

I can't think of any examples of harm, but it should be simple.

``` python
class Body1(BaseModel):
  name1: str
  
class Body2(BaseModel):
  name2: str

@app.post("/func1")
def func1(body1: Body1):
  ...

@app.post("/func2")  
def func2(body1: Body1, body2: Body2):
  ...

# more simple and explicit.
@app.post("/func3")
class Func3(BaseModel):
  body1: Body1
  body2: Body2

  def  __call__(self):
    ... 
```

> However, if it's just an extension to FastAPI, maybe it should be a separate package.

I will publish the package if there is demand.
Perhaps the sample code should be sufficient.
